tests/integration/framework/process/placement/placement.go

```diff
@@ -164,3 +175,123 @@ func (p *Placement) InitialClusterPorts() []int {
 func (p *Placement) CurrentActorsAPILevel() int {
 	return 20 // Defined in pkg/actors/internal/api_level.go
 }
+
+func (p *Placement) EstablishConn(ctx context.Context) (*grpc.ClientConn, error) {
+	return grpc.DialContext(ctx, "localhost:"+strconv.Itoa(p.port),
+		grpc.WithBlock(),
+		grpc.WithTransportCredentials(grpcinsecure.NewCredentials()),
+	)
+}
+
+func RegisterHost(t *testing.T, ctx context.Context, conn *grpc.ClientConn, msg *placementv1pb.Host, placementMessage chan any, stopCh chan struct{}) {
+	// Establish a stream and send the initial heartbeat
+	// We need to retry here because this will fail until the instance of placement (the only one) acquires leadership
+	var placementStream placementv1pb.Placement_ReportDaprStatusClient
+
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
+		client := placementv1pb.NewPlacementClient(conn)
+
+		stream, rErr := client.ReportDaprStatus(ctx)
+		//nolint:testifylint
+		if !assert.NoError(c, rErr) {
+			log.Printf("Failed to connect to placement; will retry: %v", rErr)
+			return
+		}
+
+		rErr = stream.Send(msg)
+		//nolint:testifylint
+		if !assert.NoError(c, rErr) {
+			log.Printf("Failed to send message; will retry: %v", rErr)
+			_ = stream.CloseSend()
+			return
+		}
+
+		// Receive the first message (which can't be an "update" one anyways) to ensure the connection is ready
+		_, rErr = stream.Recv()
+		//nolint:testifylint
+		if !assert.NoError(c, rErr) {
+			log.Printf("Failed to receive message; will retry: %v", rErr)
+			_ = stream.CloseSend()
+			return
+		}
+
+		placementStream = stream
+	}, time.Second*15, time.Millisecond*500)
+
+	// Send messages every second
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				placementStream.CloseSend()
+				return
+			case <-stopCh:
+				// Disconnect when there's a signal on stopCh
+				placementStream.CloseSend()
+				return
+			case <-time.After(time.Second):
+				placementStream.Send(msg)
+			}
+		}
+	}()
+
+	// Collect all API levels
+	go func() {
+		for {
+			in, rerr := placementStream.Recv()
+			if rerr != nil {
+				if ctx.Err() != nil || errors.Is(rerr, context.Canceled) || errors.Is(rerr, io.EOF) || status.Code(rerr) == codes.Canceled {
+					// Stream ended
+					placementMessage <- nil
+					return
+				}
+				placementMessage <- fmt.Errorf("error from placement: %w", rerr)
+				return
+			}
+			if in.GetOperation() == "update" {
+				placementMessage <- in.GetTables()
+			}
+		}
+	}()
+}
+
+// Expect the registration to fail with FailedPrecondition.
+func RegisterHostFailing(t *testing.T, ctx context.Context, conn *grpc.ClientConn, apiLevel int) {
```

Comment:
Ideally, this should also receive a `msg *placementv1pb.Host` argument, like `RegisterHost`, but the way this function is used for now doesn't justify the refactor, because it only varies the `apiLevel` parameter.